### PR TITLE
Fine-tune llama on the big dataset.

### PR DESCRIPTION
**Pull request by SzymonKozl**

### CODE REVIEW

**Comment by SzymonKozl** [comment]
# A note about experiments performed
## batch_size issue
I had issues when running `fine_tune_llama.py` with `per_device_train_batch_size=16`. Lowering this hyperparameter to 8 fixed the problem. Additionally, when i have increased gradient accumulation steps to 16 I had to lower batch_size even further, to 4, which suggests that the root cause here is related to the memory availability.
## optimizers
Results of mine finetunings as well as those from [PR #79](https://github.com/ZPP-MURMURAS/ZPP_Murmuras/pull/79) indicate that after first few epochs loss explodes and never reaches initial level. I have observed this anomaly when fine-tuning with optimizer from adam family or rmsprop. SGD on the other hand has no problem with it. Usually this phenomenon is observed with a gradient explosion.

I have tried lowering lr, playing with adam params, clipping the gradient, lowering the weight decay rate, changing gradient accumulation steps and running with different lr schedulers, but with little result. My experiments were far from being exhaustive, but at this moment I recommend using SGD.
## lr schedulers
From my experience the `constant` scheduler outperforms other ones for 30 epochs. This probably will change if we will run the training with a greater number of epochs.

**Comment by SzymonKozl** [comment]
Due to the suggestion from Gustaw I will include the above note in a new notebook.

File: llama_finetuning/README.md [hunk 1]
```diff
@@ -14,4 +14,25 @@ Because of that, the fine-tuning code itself does not need any requirements.
 ## Usage
 First of all, you need to create a Wandb account for logging the results. Then, you need to specify the following environment variables:
 - `WANDB_API_KEY`: set it inside the bash script
-- `HUGGING_FACE_TOKEN`: set it inside the bash script
\ No newline at end of file
+- `HUGGING_FACE_TOKEN`: set it inside the bash script
+- `TIMEOUT`: timeout (in seconds) for modal call
+- `DATASET_NAME`: dataset from hf to use
+- `WANDB_PROJECT`: name of the wandb project that will gather training metrics
```
Comment (GustawB) [inline llama_finetuning/README.md:20]: I would also update the bash script accordingly

File: llama_finetuning/fine_tune_llama.py [hunk 2]
```diff
@@ -101,18 +98,25 @@ def train_model(model, tokenizer, dataset_name, training_data, max_seq_length):
...
 
 @app.local_entrypoint()
 def main():
-    hf_token = os.getenv('HUGGING_FACE_TOKEN')
-    wandb_key = os.getenv('WANDB_KEY')
-    dataset_name = os.getenv('DATASET_NAME')
-    wrapper.remote('meta-llama/Llama-3.2-1B', hf_token, wandb_key, dataset_name)
\ No newline at end of file
+    assert (hf_token := os.getenv('HUGGING_FACE_TOKEN')) is not None
+    assert (wandb_key := os.getenv('WANDB_KEY')) is not None
```
Comment (GustawB) [inline llama_finetuning/fine_tune_llama.py:115]: mismatch with readme

**Review by GustawB — CHANGES_REQUESTED** [review]
LGTM. I think you can move all the research to the new task, you did enough stuff here anyway. One thing though, I run the fine-tuning using the per_something_batch_size = 16 for 25 epoch 4 times, with linear scheduler and had no issues. Could you try to reproduce the exact error?

File: tools/generate_llama_dataset_generic.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
+import os
```
[the remark this answers is not in the document]
Comment (SzymonKozl) [reply tools/generate_llama_dataset_generic.py:1]: done

File: llama_finetuning/README.md (hunk 1 above)
Comment (SzymonKozl) [reply llama_finetuning/README.md:20]: done

File: llama_finetuning/fine_tune_llama.py (hunk 2 above)
Comment (SzymonKozl) [reply llama_finetuning/fine_tune_llama.py:115]: fixed

**Review by GustawB — APPROVED** [review]
LGTM
